### PR TITLE
Update syn quote

### PR DESCRIPTION
**Pull request by Robbepop**
This PR updates dependencies `syn` from 0.11.x to 0.15.x and `quote` from 0.3.x to 0.6.x.

Improvements to the documentation have been made.
Also restructured and simplified some code.

### CODE REVIEW

**Comment by Robbepop** [comment]
Note that Travis CI fails because of some unstable features used in stable context. I currently do not know why this happens.

File: derive/src/lib.rs [hunk 1]
```diff
@@ -1,10 +1,12 @@
 //! Ethereum (Solidity) derivation for rust contracts (compiled to wasm or otherwise)
 #![feature(use_extern_macros)]
-#![recursion_limit="128"]
+#![recursion_limit="256"]
```
Comment (Robbepop) [inline derive/src/lib.rs]: I should have reverted this back to 128 since the 256 are not required anyway.

File: derive/src/lib.rs [hunk 2]
```diff
@@ -16,110 +18,209 @@ mod items;
...
+}
+
+/// Generates the eth abi code in case of a single provided endpoint.
+fn generate_eth_endpoint_wrapper(
+	intf: &items::Interface,
+	endpoint_name: &str
+)
+	-> proc_macro2::TokenStream
+{
+	// === REFACTORING TARGET ===
```
Comment (NikVolf) [inline derive/src/lib.rs]: what is this about?

**Comment by pepyakin** [comment]
Before continuing work on that I would really love to see CI working again.

File: derive/src/utils.rs [hunk 3]
```diff
@@ -136,6 +159,11 @@ pub fn canonical(name: &syn::Ident, method_sig: &syn::MethodSig) -> String {
 	s
 }
 
+/// Returns the Keccak hash (256-bits) of the given string slice.
+/// 
+/// # Note
+/// 
+/// The given string slice will be interpreted as bytes array.
```
Comment (pepyakin) [inline derive/src/utils.rs]: As such, maybe it'd be better if this function actually took the byte slice?

File: derive/src/utils.rs [hunk 4]
```diff
@@ -144,6 +172,12 @@ pub fn keccak(s: &str) -> H256 {
 	res
 }
 
+/// Returns the first 4 bytes of the Keccak hash (256-bits)
+/// for the given string slice.
+/// 
+/// # Note
+/// 
+/// The given string slice will be interpreted as bytes array.
```
Comment (pepyakin) [inline derive/src/utils.rs]: May I suggest renaming this function to something like `function selector`? `hash` is very ambiguous term, and [function selector](https://github.com/ethereum/wiki/wiki/Ethereum-Contract-ABI#function-selector) is more specific.

File: derive/src/lib.rs [hunk 5]
```diff
@@ -16,110 +18,209 @@ mod items;
...
-pub fn eth_abi(args: TokenStream, input: TokenStream) -> TokenStream {
-	let source = input.to_string();
-	let ast = syn::parse_item(&source).expect("Failed to parse derive input");
+pub fn eth_abi(
+	args: proc_macro::TokenStream,
+	input: proc_macro::TokenStream
+)
+	-> proc_macro::TokenStream
+{
+    let args: proc_macro2::TokenStream = args.into();
```
Comment (pepyakin) [inline derive/src/lib.rs]: Could you change spaces here to tabs?

File: derive/src/utils.rs [hunk 6]
```diff
@@ -61,63 +66,81 @@ pub fn produce_signature<T: quote::ToTokens>(
...
+		"i32"     => target.push_str("int32"),
+		"u64"     => target.push_str("uint64"),
+		"i64"     => target.push_str("int64"),
+		"U256"    => target.push_str("uint256"),
+		"H256"    => target.push_str("uint256"),
+		"Address" => target.push_str("address"),
+		"String"  => target.push_str("string"),
+		"bool"    => target.push_str("bool"),
+		"Vec"     => push_canonical_vec(target, &seg.arguments),
+		val       => panic!("[e1] Unable to handle param of type {}: not supported by abi", val)
```
Comment (pepyakin) [inline derive/src/utils.rs]: Vertical alignment isn't something that we have in our house style. Certainly not mixing spaces and tabs : )

File: derive/src/lib.rs [hunk 7]
```diff
@@ -16,110 +18,209 @@ mod items;
...
+			generate_eth_endpoint_wrapper(&intf, args.endpoint_name())
+		},
+		Some(client_name) => {
+			generate_eth_endpoint_and_client_wrapper(&intf, args.endpoint_name(), client_name)
 		}
-		len => {
-			panic!("eth_abi marco takes one or two comma-separated arguments, passed {}", len);
+	};
+
+    output.into()
```
Comment (pepyakin) [inline derive/src/lib.rs]: Could you change spaces to tabs?

File: derive/src/lib.rs (hunk 2 above)
Comment (pepyakin) [reply derive/src/lib.rs]: Would be nice to lay down such kinda of plan in to the comment.

File: derive/src/utils.rs (hunk 4 above)
Comment (Robbepop) [reply derive/src/utils.rs]: Yeah I would also think that this is a better name for it. I will rename it.

File: derive/src/lib.rs (hunk 5 above)
Comment (Robbepop) [reply derive/src/lib.rs]: sure

File: derive/src/utils.rs (hunk 6 above)
Comment (Robbepop) [reply derive/src/utils.rs]: I can understand that. I will revert the extra vertical alignment.

File: derive/src/lib.rs (hunk 7 above)
Comment (Robbepop) [reply derive/src/lib.rs]: I think I have to inspect settings of my rust-fmt.

File: derive/src/lib.rs (hunk 2 above)
Comment (Robbepop) [reply derive/src/lib.rs]: True! Will do that!

**Comment by Robbepop** [comment]
> Before continuing work on that I would really love to see CI working again.

Master doesn't work either as far as I have tested. But it should be when adding `+nightly` to `cargo` command. I will test that.

File: derive/src/utils.rs (hunk 4 above)
Comment (Robbepop) [reply derive/src/utils.rs]: Renamed it, resolved.

File: derive/src/utils.rs (hunk 3 above)
Comment (Robbepop) [reply derive/src/utils.rs]: Good point! Did so in the latest set of commits.

File: derive/src/lib.rs (hunk 2 above)
Comment (Robbepop) [reply derive/src/lib.rs]: Done

File: derive/src/lib.rs [hunk 8]
```diff
@@ -16,110 +18,209 @@ mod items;
...
+	}
+}
+
+/// Parses the given token stream as arguments for the `eth_abi` attribute macro.
+fn parse_args(args: proc_macro2::TokenStream) -> Args {
+	let args = parse_args_to_vec(args);
+
+	assert!(1 <= args.len() && args.len() <= 2,
+		"[err01]: Expect one argument for endpoint name and an optional argument for client name.");
+	println!("eth_abi::args = {:?}", args);
```
[the remark this answers is not in the document]
Comment (Robbepop) [reply derive/src/lib.rs]: No it was a debugging artifact and got already removed, thanks! :)

File: derive/src/items.rs [hunk 9]
```diff
@@ -107,55 +135,60 @@ fn into_signature(ident: syn::Ident, method_sig: syn::MethodSig, is_constant: bo
...
+		};
+		Item::Event(event)
+	}
+
+	fn signature_from_trait_item(method_trait_item: syn::TraitItemMethod) -> Self {
+		let constant = has_attribute(&method_trait_item.attrs, "constant");
+		let payable = has_attribute(&method_trait_item.attrs, "payable");
+		assert!(!(constant && payable),
+			format!("Method {} cannot be constant and payable at the same time", method_trait_item.sig.ident.to_string()
+		));
```
Comment (fckt) [inline derive/src/items.rs]: what if we move  `method_trait_item.sig.ident.to_string()` on the next line? E.g.:

```
	assert!(
		!(constant && payable),
		format!(
			"Method {} cannot be constant and payable at the same time",
			method_trait_item.sig.ident.to_string()
		)
	);
```

**Comment by debris** [comment]
I see that there are no unit tests for generated token stream. As debugging `syn` && `quote` is extremely painful I suggest adding some tests for generated token streams 😉 

e.g. we have something like this in `ethabi`

https://github.com/paritytech/ethabi/blob/523cb5bcb41607737e52da051d46e223cc54b212/derive/src/event.rs#L322-L344

```rust
	#[test]
	fn test_log_with_one_field() {
		let ethabi_event = ethabi::Event {
			name: "one".into(),
			inputs: vec![ethabi::EventParam {
				name: "foo".into(),
				kind: ethabi::ParamType::Address,
				indexed: false
			}],
			anonymous: false,
		};

		let e = Event::from(&ethabi_event);

		let expected = quote! {
			#[derive(Debug, Clone, PartialEq)]
			pub struct One {
				pub foo: ethabi::Address
			}
		};

		assert_eq!(expected.to_string(), e.generate_log().to_string());
	}
```

**Comment by Robbepop** [comment]
> I see that there are no unit tests for generated token stream. As debugging `syn` && `quote` is extremely painful I suggest adding some tests for generated token streams wink
> 
> e.g. we have something like this in `ethabi`
> 
> https://github.com/paritytech/ethabi/blob/523cb5bcb41607737e52da051d46e223cc54b212/derive/src/event.rs#L322-L344
> 
> ```rust
> 	#[test]
> 	fn test_log_with_one_field() {
> 		let ethabi_event = ethabi::Event {
> 			name: "one".into(),
> 			inputs: vec![ethabi::EventParam {
> 				name: "foo".into(),
> 				kind: ethabi::ParamType::Address,
> 				indexed: false
> 			}],
> 			anonymous: false,
> 		};
> 
> 		let e = Event::from(&ethabi_event);
> 
> 		let expected = quote! {
> 			#[derive(Debug, Clone, PartialEq)]
> 			pub struct One {
> 				pub foo: ethabi::Address
> 			}
> 		};
> 
> 		assert_eq!(expected.to_string(), e.generate_log().to_string());
> 	}
> ```

Thank you very much for hinting that to me!
I actually didn't know about this testing functionality but it sounds awesome.

Maybe we should implement the tests in another PR as this is already too big.

File: derive/src/items.rs (hunk 9 above)
Comment (Robbepop) [reply derive/src/items.rs]: Sure! :)

**Comment by Robbepop** [comment]
@fckt I hope I resolved the formatting issues to your satisfaction in the most recent commits.

Besides that I have added some proper documentation for the items defined in `items.rs` as well.

File: derive/src/items.rs (hunk 9 above)
Comment (Robbepop) [reply derive/src/items.rs]: done

**Comment by Robbepop** [comment]
Thank you all for the reviews and feedback! :)

**Comment by Robbepop** [comment]
@NikVolf I think we can also delete the `syn-upgrade` branch.

**Pull request merged**